File: top-interview-questions-medium/Sorting and Searching/top-k-frequent-elements/347.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
  let map = {}
  for (let i of nums) {
    map[i] = map[i] !== undefined ? map[i] + 1 : 1
  }

  let list = []
  for (let i in map) {
    if (list[map[i]] !== undefined) {
      list[map[i]].push(Number(i))
    } else {
      list[map[i]] = [Number(i)]
    }
  }

  let res = []
  while (res.length < k) {
    let v = list.pop()
    if (v == undefined) continue
    res = res.concat(v)
  }
  return res
}

File: top-interview-questions-medium/Sorting and Searching/top-k-frequent-elements/347.py
class Solution:
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        map = {}
        for i in nums:
            map[i] = map[i] + 1 if i in map else 1

        li = [None] * (len(nums) + 1)
        for key, value in map.items():
            if li[value] is not None:
                li[value].append(k)
            else:
                li[value] = [key]

        print(li)

        res = []
        while len(res) < k:
            last = li.pop()
            if last is None: continue
            res += last

        return res
